src/lib/FreeC/Pass/CompletePatternPass.hs

```diff
@@ -0,0 +1,146 @@
+--
+--   should not pass the check because the @case@ expression has two alternatives
+--   for the constructor @Just@.
+--
+--   == Example 3
+--   The following declaration where the scrutinee is a function
+--
+--   > case_id = case \x -> x  of
+--
+--   should not pass the check because the type of the scrutinee.
+--
+--   = Specification
+--
+--   == Preconditions
+--
+--   The type of all checked expressions has to be annotated.
+--   The Environment has to contain the names of all constructors for
+--   all used data types.
+--   Additionally, the environment should contain entries for all used type
+--   synonyms.
+--
+--   == Translation
+--
+--   This pass only performs a check and therefore does not change the module.
+--   == Postconditions
+--
+--   All case expressions are guaranteed to have complete pattern matching.
+--
+--   == Error cases
+--
+--   A fatal error is reported if an incomplete case expression is found.
+
+module FreeC.Pass.CompletePatternPass
+  ( completePatternPass
+  , checkPatternFuncDecl
+  )
+where
+
+import           Control.Monad                  ( unless )
+import           Data.Maybe                     ( fromJust )
+
+import           FreeC.Environment
+import           FreeC.Environment.Entry
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.IR.SrcSpan
+import           FreeC.IR.TypeSynExpansion
+import           FreeC.Monad.Converter
+import           FreeC.Monad.Reporter
+import           FreeC.Pass
+import           FreeC.Pretty                   ( showPretty )
+
+-- | Checks that all functions of a given module have complete pattern matching.
+--
+--   The pattern matching for a function is complete if for each @case@
+--   expression there exists exactly one case alternative for each constructor
+--   of the corresponding type.
+completePatternPass :: Pass IR.Module
+completePatternPass ast = do
+  mapM_ checkPatternFuncDecl (IR.modFuncDecls ast)
+  return ast
+
+-- | Checks a function declaration for incomplete patterns.
+checkPatternFuncDecl :: IR.FuncDecl -> Converter ()
+checkPatternFuncDecl funcDecl = checkPatternExpr (IR.funcDeclRhs funcDecl)
+ where
+  -- | Checks an expression for incomplete patterns.
+  checkPatternExpr :: IR.Expr -> Converter ()
+  checkPatternExpr (IR.Case srcSpan exprScrutinee exprAlts _) = do
+    checkPatternExpr exprScrutinee
+    mapM_ (checkPatternExpr . IR.altRhs) exprAlts
+    let tau = fromJust $ IR.exprType exprScrutinee -- is safe beacause all types are annotated
+    tau' <- expandAllTypeSynonyms tau
+    case getTypeConName tau' of
+      Nothing       -> failedPatternCheck srcSpan
+      Just typeName -> do
+        maybeEntry <- inEnv $ lookupEntry IR.TypeScope typeName
+        let altConNames = map (IR.conPatName . IR.altConPat) exprAlts
+        case maybeEntry of
+          Just entry | isDataEntry entry ->
+            performCheck (entryConsNames entry) altConNames srcSpan
+          _ -> failedPatternCheck srcSpan
```

Comment:
The error message printed by `failedPatternCheck` does not describe the actual error in this case very well. The catch all case applies if `maybeEntry` is `Nothing` (i.e., `typeName` is not in the environment) or the entry is not a `DataEntry`. The case that the entry is missing can be handled using [`lookupEntryOrFail`](https://freeproving.github.io/free-compiler/docs/master/FreeC-Environment-LookupOrFail.html#v:lookupEntryOrFail). That the entry is a `DataEntry` should already be guaranteed by `expandAllTypeSynonyms` (i.e., it cannot be a `TypeSynEntry`) and `getTypeConName` (i.e., it cannot be a `TypeVarEntry`). Thus, I think you can omit the second check. However, you should state the assumption in a comment.